File: clinical-schema/python/src/dbbest_clinical_schema/parsers/hl7_composition.py
"""Mirror js/src/parsers/hl7-composition.js."""
from __future__ import annotations

from typing import Any

from ..constants import ACTIVE_STATUS_ACTIVE, InOut, ValType, VITAL_STATUS_ALIVE
from ..models._base import iso_date, iso_now


def parse_hl7_composition(envelope: dict) -> dict[str, list[dict]]:
    if not isinstance(envelope, dict):
        raise TypeError("parse_hl7_composition: envelope must be a dict")
    cda = envelope if envelope.get("resourceType") == "Composition" else envelope.get("cda")
    if not cda or cda.get("resourceType") != "Composition":
        raise ValueError("parse_hl7_composition: expected resourceType='Composition'")
    sections = cda.get("section")
    if not isinstance(sections, list):
        raise ValueError("parse_hl7_composition: Composition.section must be a list")
    return _extract_data_from_sections(sections)


def _extract_data_from_sections(sections: list[dict]) -> dict:
    patients: list[dict] = []
    visits: list[dict] = []
    observations: list[dict] = []
    patient_map: dict[str, dict] = {}
    visit_map: dict[int, dict] = {}
    patient_counter = 1
    visit_counter = 1
    observation_counter = 1

    for section in sections:
        entries = section.get("entry")
        if not isinstance(entries, list):
            continue
        title = section.get("title")
        if title == "Patient Information":
            _extract_patients(entries, patients, patient_map, patient_counter)
            patient_counter += len(entries)
        elif isinstance(title, str) and title.startswith("Visit "):
            _extract_visit(section, visits, visit_map, visit_counter, patient_map)
            visit_counter += 1
        else:
            _extract_observations(section, observations, observation_counter, patient_map, visit_map)
            observation_counter += len(entries)

    return {"patients": patients, "visits": visits, "observations": observations}


def _extract_patients(entries: list[dict], patients: list[dict], patient_map: dict, start: int) -> None:
    current = None
    num = start
    for entry in entries:
        title = entry.get("title", "")
        if title.startswith("Patient: "):
            if current:
                patients.append(current)
                patient_map[current["PATIENT_CD"]] = current
            current = {
                "PATIENT_NUM": num,
                "PATIENT_CD": entry.get("value"),
                "VITAL_STATUS_CD": VITAL_STATUS_ALIVE,
                "BIRTH_DATE": None, "DEATH_DATE": None, "AGE_IN_YEARS": None,
                "SEX_CD": None, "LANGUAGE_CD": None, "RACE_CD": None,
                "MARITAL_STATUS_CD": None, "RELIGION_CD": None,
                "STATECITYZIP_PATH": None, "PATIENT_BLOB": None,
                "UPDATE_DATE": iso_date(), "DOWNLOAD_DATE": None,
                "IMPORT_DATE": iso_now(), "SOURCESYSTEM_CD": "HL7_IMPORT",
                "UPLOAD_ID": 1, "CREATED_AT": iso_now(), "UPDATED_AT": iso_now(),
            }
            num += 1
        elif current:
            if title == "Gender":         current["SEX_CD"] = entry.get("value")
            elif title == "Age":           current["AGE_IN_YEARS"] = entry.get("value")
            elif title == "Date of birth": current["BIRTH_DATE"] = entry.get("value")
    if current:
        patients.append(current)
        patient_map[current["PATIENT_CD"]] = current


def _extract_visit(section: dict, visits: list[dict], visit_map: dict, num: int, patient_map: dict) -> None:
    visit_date: Any = None
    location: Any = None
    for entry in section.get("entry") or []:
        if entry.get("title") == "Visit Date": visit_date = entry.get("value")
        elif entry.get("title") == "Location":  location = entry.get("value")

    patient_ids = list(patient_map.keys())
    if not patient_ids:
        return
    pid = patient_ids[0] if num <= 2 else (patient_ids[1] if len(patient_ids) > 1 else patient_ids[0])
    patient = patient_map.get(pid)
    if not patient:
        return
    visit = {
        "ENCOUNTER_NUM": num,
        "PATIENT_NUM": patient["PATIENT_NUM"],
        "ACTIVE_STATUS_CD": ACTIVE_STATUS_ACTIVE,
        "START_DATE": visit_date or iso_date(),
        "END_DATE": None,
        "INOUT_CD": _determine_visit_type(location),
        "LOCATION_CD": location or "HL7_IMPORT",
        "VISIT_BLOB": None,
        "UPDATE_DATE": None, "DOWNLOAD_DATE": None, "IMPORT_DATE": None,
        "SOURCESYSTEM_CD": "HL7_IMPORT", "UPLOAD_ID": 1, "CREATED_AT": iso_date(),
    }
    visits.append(visit)
    visit_map[num] = visit


def _extract_observations(section: dict, observations: list[dict], start: int, patient_map: dict, visit_map: dict) -> None:
    num = start
    for entry in section.get("entry") or []:
        title = entry.get("title")
        if title in ("Visit Date", "Location"):
            continue
        value = entry.get("value")
        if isinstance(value, bool):
            valtype, tval, nval = ValType.TEXT.value, str(value), None
        elif isinstance(value, (int, float)):
            valtype, tval, nval = ValType.NUMERIC.value, None, float(value)
        elif isinstance(value, str):
            valtype, tval, nval = ValType.TEXT.value, value, None
        elif value is None:
            valtype, tval, nval = ValType.TEXT.value, None, None
        else:
            valtype, tval, nval = ValType.TEXT.value, str(value), None
        patient_ids = list(patient_map.keys())
        if not patient_ids:
            continue
        patient = patient_map[patient_ids[0]]
        visit_ids = list(visit_map.keys())
        observations.append({
            "OBSERVATION_ID": num,
            "ENCOUNTER_NUM": visit_ids[0] if visit_ids else None,
            "PATIENT_NUM": patient["PATIENT_NUM"],
            "CATEGORY_CHAR": None,
            "CONCEPT_CD": title,
            "PROVIDER_ID": None,
            "START_DATE": iso_date(),
            "INSTANCE_NUM": 1,
            "VALTYPE_CD": valtype, "TVAL_CHAR": tval, "NVAL_NUM": nval,
            "VALUEFLAG_CD": None, "QUANTITY_NUM": None, "UNIT_CD": None,
            "END_DATE": None, "LOCATION_CD": None, "CONFIDENCE_NUM": None,
            "OBSERVATION_BLOB": None,
            "UPDATE_DATE": None, "DOWNLOAD_DATE": None, "IMPORT_DATE": None,
            "SOURCESYSTEM_CD": "HL7_IMPORT", "UPLOAD_ID": 1, "CREATED_AT": iso_date(),
        })
        num += 1


def _determine_visit_type(location: Any) -> str:
    if not location:
        return InOut.OUTPATIENT.value
    lower = str(location).lower()
    if "emergency" in lower or "notaufnahme" in lower:
        return InOut.EMERGENCY.value
    if "hospital" in lower or "inpatient" in lower or "station" in lower:
        return InOut.INPATIENT.value
    if "clinic" in lower or "ambulanz" in lower:
        return InOut.OUTPATIENT.value
    return InOut.OUTPATIENT.value

File: clinical-schema/js/src/parsers/hl7-composition.js
import { VALTYPE, VITAL_STATUS, ACTIVE_STATUS, INOUT } from '../constants.js'
import { isoNow, isoDate } from '../builders/util.js'

/**
 * Parse an HL7 FHIR Composition envelope into normalized record arrays.
 *
 * Accepts either { cda: { resourceType: 'Composition', section: [...] } }
 * or a bare Composition resource.
 *
 * Behavior matches dbBEST `ImportHl7Service.extract*` (src/core/services/
 * imports/import-hl7-service.js, lines 140–394).
 */
export function parseHl7Composition(envelope) {
  if (!envelope || typeof envelope !== 'object') {
    throw new Error('parseHl7Composition: envelope must be an object')
  }
  const cda = envelope.resourceType === 'Composition' ? envelope : envelope.cda
  if (!cda || cda.resourceType !== 'Composition') {
    throw new Error('parseHl7Composition: expected resourceType="Composition"')
  }
  if (!Array.isArray(cda.section)) {
    throw new Error('parseHl7Composition: Composition.section must be an array')
  }
  return extractDataFromSections(cda.section)
}

function extractDataFromSections(sections) {
  const patients = []
  const visits = []
  const observations = []
  const patientMap = new Map()
  const visitMap = new Map()
  let patientCounter = 1
  let visitCounter = 1
  let observationCounter = 1

  for (const section of sections) {
    if (!Array.isArray(section.entry)) continue

    if (section.title === 'Patient Information') {
      extractPatientsFromSection(section.entry, patients, patientMap, patientCounter)
      patientCounter += section.entry.length
    } else if (section.title?.startsWith('Visit ')) {
      extractVisitFromSection(section, visits, visitMap, visitCounter, patientMap)
      visitCounter++
    } else {
      extractObservationsFromSection(section, observations, observationCounter, patientMap, visitMap)
      observationCounter += section.entry.length
    }
  }
  return { patients, visits, observations }
}

function extractPatientsFromSection(entries, patients, patientMap, startCounter) {
  let current = null
  let patientNum = startCounter
  for (const entry of entries) {
    if (entry.title?.startsWith('Patient: ')) {
      if (current) {
        patients.push(current)
        patientMap.set(current.PATIENT_CD, current)
      }
      current = {
        PATIENT_NUM: patientNum++,
        PATIENT_CD: entry.value,
        VITAL_STATUS_CD: VITAL_STATUS.ALIVE,
        BIRTH_DATE: null,
        DEATH_DATE: null,
        AGE_IN_YEARS: null,
        SEX_CD: null,
        LANGUAGE_CD: null,
        RACE_CD: null,
        MARITAL_STATUS_CD: null,
        RELIGION_CD: null,
        STATECITYZIP_PATH: null,
        PATIENT_BLOB: null,
        UPDATE_DATE: isoDate(),
        DOWNLOAD_DATE: null,
        IMPORT_DATE: isoNow(),
        SOURCESYSTEM_CD: 'HL7_IMPORT',
        UPLOAD_ID: 1,
        CREATED_AT: isoNow(),
        UPDATED_AT: isoNow(),
      }
    } else if (current && entry.title === 'Gender') {
      current.SEX_CD = entry.value
    } else if (current && entry.title === 'Age') {
      current.AGE_IN_YEARS = entry.value
    } else if (current && entry.title === 'Date of birth') {
      current.BIRTH_DATE = entry.value
    }
  }
  if (current) {
    patients.push(current)
    patientMap.set(current.PATIENT_CD, current)
  }
}

function extractVisitFromSection(section, visits, visitMap, visitNum, patientMap) {
  let visitDate = null
  let location = null
  for (const entry of section.entry) {
    if (entry.title === 'Visit Date') visitDate = entry.value
    else if (entry.title === 'Location') location = entry.value
  }
  // Heuristic patient assignment (mirrors dbBEST behavior): first 2 visits to first patient, rest to second
  const patientIds = Array.from(patientMap.keys())
  const patientId = patientIds[visitNum <= 2 ? 0 : 1] ?? patientIds[0]
  const patient = patientMap.get(patientId)
  if (!patient) return
  const visit = {
    ENCOUNTER_NUM: visitNum,
    PATIENT_NUM: patient.PATIENT_NUM,
    ACTIVE_STATUS_CD: ACTIVE_STATUS.ACTIVE,
    START_DATE: visitDate || isoDate(),
    END_DATE: null,
    INOUT_CD: determineVisitType(location),
    LOCATION_CD: location || 'HL7_IMPORT',
    VISIT_BLOB: null,
    UPDATE_DATE: null,
    DOWNLOAD_DATE: null,
    IMPORT_DATE: null,
    SOURCESYSTEM_CD: 'HL7_IMPORT',
    UPLOAD_ID: 1,
    CREATED_AT: isoDate(),
  }
  visits.push(visit)
  visitMap.set(visitNum, visit)
}

function extractObservationsFromSection(section, observations, startCounter, patientMap, visitMap) {
  let observationNum = startCounter
  for (const entry of section.entry) {
    if (entry.title === 'Visit Date' || entry.title === 'Location') continue
    let valtypeCd = VALTYPE.TEXT
    let tvalChar = null
    let nvalNum = null
    if (typeof entry.value === 'number') {
      valtypeCd = VALTYPE.NUMERIC
      nvalNum = entry.value
    } else if (typeof entry.value === 'string') {
      valtypeCd = VALTYPE.TEXT
      tvalChar = entry.value
    } else if (entry.value !== undefined && entry.value !== null) {
      valtypeCd = VALTYPE.TEXT
      tvalChar = String(entry.value)
    }
    const patientIds = Array.from(patientMap.keys())
    const patient = patientMap.get(patientIds[0])
    const visitIds = Array.from(visitMap.keys())
    const visitId = visitIds[0]
    if (!patient) continue
    observations.push({
      OBSERVATION_ID: observationNum++,
      ENCOUNTER_NUM: visitId,
      PATIENT_NUM: patient.PATIENT_NUM,
      CATEGORY_CHAR: null,
      CONCEPT_CD: entry.title,
      PROVIDER_ID: null,
      START_DATE: isoDate(),
      INSTANCE_NUM: 1,
      VALTYPE_CD: valtypeCd,
      TVAL_CHAR: tvalChar,
      NVAL_NUM: nvalNum,
      VALUEFLAG_CD: null,
      QUANTITY_NUM: null,
      UNIT_CD: null,
      END_DATE: null,
      LOCATION_CD: null,
      CONFIDENCE_NUM: null,
      OBSERVATION_BLOB: null,
      UPDATE_DATE: null,
      DOWNLOAD_DATE: null,
      IMPORT_DATE: null,
      SOURCESYSTEM_CD: 'HL7_IMPORT',
      UPLOAD_ID: 1,
      CREATED_AT: isoDate(),
    })
  }
}

function determineVisitType(location) {
  if (!location) return INOUT.OUTPATIENT
  const lower = String(location).toLowerCase()
  if (lower.includes('emergency') || lower.includes('notaufnahme')) return INOUT.EMERGENCY
  if (lower.includes('hospital') || lower.includes('inpatient') || lower.includes('station')) return INOUT.INPATIENT
  if (lower.includes('clinic') || lower.includes('ambulanz')) return INOUT.OUTPATIENT
  return INOUT.OUTPATIENT
}
